File: code/binding/development/tools/cast/src/c++/cast/architecture/ManagedProcess.cpp
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

#include "ManagedProcess.hpp"
#include <sstream>
#include <exception>

using namespace std;

namespace cast {

  ManagedProcess::ManagedProcess(const string &_id) :
    WorkingMemoryAttachedComponent(_id),
    WorkingMemoryWriterProcess(_id),
    WorkingMemoryReaderProcess(_id),
    m_notifications(0),
    m_pProposalInputToTaskManager(NULL),
    m_pRetractionInputToTaskManager(NULL),
    m_pProcessingResultToGoalManager(NULL),
    m_pTaskDescriptionRegistration(NULL),
    m_pProposedTasks(new IPTaskMap()),
    m_taskCounter(0),
    m_responseRunnable(NULL)
  {
    
    m_pTaskNotificationCondition 
      = new omni_condition(&m_taskNotificationMutex);

  }

  ManagedProcess::~ManagedProcess() {


    QueuedDataThread<cdl::TaskManagementResult>::cleanup(m_responseRunnable);

    m_taskNotificationMutex.lock();
    //dummy value to prevent assert on exit
    m_notifications = 100;
    m_pTaskNotificationCondition->broadcast();
    m_taskNotificationMutex.unlock();
    delete m_pTaskNotificationCondition;

    //TODO might need to empty it too...
    delete m_pProposedTasks;
  }

  void ManagedProcess::stop() {
    //FrameworkProcess::
    WorkingMemoryReaderProcess::stop();

    //stop threads
    if(m_responseRunnable != NULL) {
      m_responseRunnable->stop();
    }


    //release sleeping threads
    m_taskNotificationMutex.lock();
    m_pTaskNotificationCondition->broadcast();
    m_taskNotificationMutex.unlock();
  }

  void ManagedProcess::waitForNotifications() {

    m_taskNotificationMutex.lock();    
    //if there are no outstanding notifications, then block
    if(m_notifications == 0) {
      //println("blocking for new");
      m_pTaskNotificationCondition->wait();
      //the count will have been incremented in the same action as the broadcast
      m_notifications--;
      assert(m_notifications >= 0);

    }
    //else decrement the coutn
    else {
      m_notifications--;
      assert(m_notifications >= 0);
    }
    m_taskNotificationMutex.unlock();	            
    

  }


  void ManagedProcess::setPushConnector(const string & _connectionID, 
					PushConnectorOut<cdl::InformationProcessingTask> * _pOut) {
    m_pProposalInputToTaskManager = _pOut;
  }

  void ManagedProcess::setPushConnector(const string & _connectionID, 
					PushConnectorOut<std::string> * _pOut) {
    m_pRetractionInputToTaskManager = _pOut;
  }

  void ManagedProcess::setPushConnector(const string & _connectionID, 
					PushConnectorOut<cdl::TaskResult> * _pOut) {
    m_pProcessingResultToGoalManager = _pOut;
  }

  void ManagedProcess::setPushConnector(const string & _connectionID, 
					PushConnectorOut<cdl::TaskDescriptionList> * _pOut) {
    m_pTaskDescriptionRegistration = _pOut;
  }




  void ManagedProcess::receivePushData(FrameworkLocalData<cdl::TaskManagementResult> *_pData) {


    //protect access to data structures
    m_gdpMutex.lock();


    try {


      if(m_status == STATUS_RUN) {

	//check that the decision is for this process
	string taskID(_pData->data()->m_id);
	IPTaskMap::iterator i = m_pProposedTasks->find(taskID);
	
	//printfln("received decision!!, %s", taskID.c_str());

	//if the task is one we proposed
	if(i != m_pProposedTasks->end()) {
	  
	  if(m_responseRunnable == NULL) { 	    
	    m_responseRunnable = new GMResponseHandler(this);
	    m_responseRunnable->start();
	  }
	
	  //copy ;)
	  //	  println("queued!!");
	  m_responseRunnable->queue(*_pData->data());
    
	} else {
	  //not for us! ... 
	  //println("shit!!");
	}
      }


    }
    catch (exception& e)
      {
	println("exception details follow...");
	cerr << e.what() << endl;
      }
    catch (...) {
      println("exception details follow...");
      cerr << "default exception" << endl;
    }

    m_gdpMutex.unlock();
  
    // data needs to be deleted whatever happens
    delete _pData;


  }


  string ManagedProcess::newTaskID() {
    std::ostringstream o;
    o << "IPG:";
    o << getProcessIdentifier();
    o << ":";
    o << m_taskCounter;

    m_taskCounter++;

    return o.str();    
  }


  void ManagedProcess::proposeInformationProcessingTask(const string &_taskID, const string &_taskName) {

    if(m_pProcessingResultToGoalManager) {

      logTaskProposed(_taskID,_taskName);

      //create the goal structure
      cdl::InformationProcessingTask *pIPT 
	= new cdl::InformationProcessingTask();
      pIPT->m_id = CORBA::string_dup(_taskID.c_str());
      pIPT->m_taskName =  CORBA::string_dup(_taskName.c_str());

      //store it as proposed
      (*m_pProposedTasks)[_taskID] = pIPT;

      //COPY (not point) data into local data object... as it will
      //get deleted by the sending process
      FrameworkLocalData<cdl::InformationProcessingTask> *pFLD 
	= new FrameworkLocalData<cdl::InformationProcessingTask>(getProcessIdentifier(),*pIPT);
    
      // push task to s-a goal manager
      m_pProposalInputToTaskManager->push(pFLD);
      
    }
    else {
      throw(SubarchitectureProcessException(__HERE__,"Push connection not to sa goal manager for proposal not set."));
    }

  }

  void ManagedProcess::retractInformationProcessingTask(const string &_taskID) {

    if(m_pRetractionInputToTaskManager) {

      FrameworkLocalData<std::string> *pFLD 
	= new FrameworkLocalData<std::string>(getProcessIdentifier(),new string(_taskID));
    
      // push task to s-a goal manager
      m_pRetractionInputToTaskManager->push(pFLD);
      
    }
    else {
      throw(SubarchitectureProcessException(__HERE__,"Push connection not to sa goal manager for retraction not set."));
    }

  }

  void ManagedProcess::taskComplete(const string &_goalID, 
				    const cdl::TaskOutcome &_outcome) {

    if (m_pProcessingResultToGoalManager) {

      logTaskComplete(_goalID);

      IPTaskMap::iterator i = m_pProposedTasks->find(_goalID);
      if(i != m_pProposedTasks->end()) {
	m_pProposedTasks->erase(i);
      }


      cdl::TaskResult *pTR = new cdl::TaskResult();
      pTR->m_id = CORBA::string_dup(_goalID.c_str());
      pTR->m_outcome = _outcome;

      //use our memory for the data, but the FLD class will delete when
      //sending is complete
      FrameworkLocalData<cdl::TaskResult> *pFLD 
	= new FrameworkLocalData<cdl::TaskResult>(getProcessIdentifier(),
						  pTR);
      m_pProcessingResultToGoalManager->push(pFLD);
    }
    else {
      throw(SubarchitectureProcessException(__HERE__,"Goal processing result connection to goal manager has not been set up."));
    }
  }

  void ManagedProcess::registerTaskDescriptions(cdl::TaskDescriptionList * _pTaskList) {

    if (m_pTaskDescriptionRegistration) {
    
      FrameworkLocalData<cdl::TaskDescriptionList> *pFLD 
	= new FrameworkLocalData<cdl::TaskDescriptionList>(getProcessIdentifier(),
							   _pTaskList);
      m_pTaskDescriptionRegistration->push(pFLD);
    }
    else {
      throw(SubarchitectureProcessException(__HERE__,"Task registration connection to goal manager has not been set up."));
    }

  }

  void 
  ManagedProcess::GMResponseHandler::nextInQueue(cdl::TaskManagementResult & _data) {
    
    if(_data.m_decision != cdl::GOAL_WAITING) {
      
      std::string taskID(_data.m_id);
      m_pGDP->lockProcess();
	  
      try {

	if(_data.m_decision == cdl::GOAL_ADOPTED) {
	  //m_pGDP->logTaskAdopted(taskID);
	  m_pGDP->taskAdopted(taskID);	    
	}
	else if(_data.m_decision == cdl::GOAL_REJECTED) {
	  m_pGDP->taskRejected(taskID);	    
	}	  
      }	  
      catch(const DoesNotExistOnWMException &e) {
	m_pGDP->println("ManagedProcess::taskAdopted DoesNotExistOnWMException caught");
	m_pGDP->println("The task involved in this error was: " + taskID);	  		
	m_pGDP->println("The address involved in this error was: " + 
			string(e.address().m_id) + " in " + 
			string(e.address().m_subarchitecture));	  		
	m_pGDP->println(e.what());
	std::abort();
      }
      catch(const CASTException &e) {
	m_pGDP->println("ManagedProcess::taskAdopted DoesNotExistOnWMException caught");
	m_pGDP->println("The task involved in this error was: " + taskID);	  		
	m_pGDP->println(e.what());
	std::abort();
      }
      catch(const BALTException &e) {
	m_pGDP->println("ManagedProcess::taskAdopted BALTException caught");
	m_pGDP->println("The task involved in this error was: " + taskID);	  		
	m_pGDP->println(e.what());
	std::abort();
      }
      catch(const CORBA::BAD_PARAM &e) {
	m_pGDP->println("ManagedProcess::taskAdopted CORBA::BAD_PARAM caught");		
	m_pGDP->println("The task involved in this error was: " + taskID);	  		
	m_pGDP->println(e._name());
	std::abort();
      }
      catch(const CORBA::Exception &e) {
	m_pGDP->println("ManagedProcess::taskAdopted CORBA::Exception caught");
	m_pGDP->println("The task involved in this error was: " + taskID);	  		
	m_pGDP->println(e._name());
	std::abort();
      }
      catch(const std::exception &e) {
	m_pGDP->println("ManagedProcess::taskAdopted std::exception caught");
	m_pGDP->println("The task involved in this error was: " + taskID);	  		
	m_pGDP->println(e.what());
	std::abort();
      }
      catch(...) {
	m_pGDP->println("ManagedProcess::taskAdopted unknown exception caught");
	m_pGDP->println("The task involved in this error was: " + taskID);	  		
	std::abort();
      }
	    
	    
      m_pGDP->unlockProcess();
	 
      //wake waiting threads
      m_pGDP->m_taskNotificationMutex.lock();
      //increment notification count
      m_pGDP->m_notifications++;    
      m_pGDP->m_pTaskNotificationCondition->broadcast();
      m_pGDP->m_taskNotificationMutex.unlock();	    	    
    }
  }



} //namespace cast 

File: binding/development/tools/cast/src/java/cast/architecture/subarchitecture/ManagedProcess.java
/*
 * CAST - The CoSy Architecture Schema Toolkit
 *
 * Copyright (C) 2006-2007 Nick Hawes
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation; either version 2.1 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301 USA
 *
 */

/**
 * 
 */
package cast.architecture.subarchitecture;

import java.util.Hashtable;

import balt.core.connectors.ControlledRunnable;
import balt.core.connectors.push.nonprimitive.interfaces.StringPushInterface.StringPushConnectorOut;
import balt.core.connectors.push.nonprimitive.interfaces.StringPushInterface.StringPushSender;
import cast.architecture.abstr.WorkingMemoryReaderWriterProcess;
import cast.cdl.InformationProcessingTask;
import cast.cdl.TaskDescription;
import cast.cdl.TaskManagementDecision;
import cast.cdl.TaskManagementResult;
import cast.cdl.TaskOutcome;
import cast.cdl.TaskResult;
import cast.cdl.ui.ComponentEventType;
import cast.core.CASTUtils;
import cast.core.components.QueuedDataRunnable;
import cast.core.interfaces.GMRPushInterface.GMRPushReceiver;
import cast.core.interfaces.GPRPushInterface.GPRPushConnectorOut;
import cast.core.interfaces.GPRPushInterface.GPRPushSender;
import cast.core.interfaces.IPGPushInterface.IPGPushConnectorOut;
import cast.core.interfaces.IPGPushInterface.IPGPushSender;
import cast.core.interfaces.TDLPushInterface.TDLPushConnectorOut;
import cast.core.interfaces.TDLPushInterface.TDLPushSender;

/**
 * An abstract class to represent a component in a subarchitecture that can read
 * and write to a working memory, and has its operations controlled by a task
 * manager.
 * 
 * @author nah
 */
public abstract class ManagedProcess extends WorkingMemoryReaderWriterProcess
		implements
			IPGPushSender,
			GMRPushReceiver,
			GPRPushSender,
			StringPushSender,
			TDLPushSender {

	/**
	 * Inner class used to forward the responses of the task manager on to the
	 * subclass.
	 * 
	 * @author nah
	 */
	private static class GMResponseHandler
			extends
				QueuedDataRunnable<TaskManagementResult> {

		private ManagedProcess m_gdp;

		/**
		 * @param _gdp
		 * @param _goalDecision
		 * @param _goalID
		 */
		public GMResponseHandler(ManagedProcess _gdp) {
			super();
			m_gdp = _gdp;
		}

		@Override
		protected void nextInQueue(TaskManagementResult _data) {
			m_gdp.lockProcess();
			m_gdp.handleTaskManagerResponse(_data.m_id, _data.m_decision);
			m_gdp.unlockProcess();
		}

	}

	/**
	 * Connection to the task manager.
	 */
	private GPRPushConnectorOut m_processingResultToGoalManager;

	/**
	 * Connection to the task manager.
	 */
	private TDLPushConnectorOut m_taskDescriptionRegistration;

	private Object m_taskNotificationLock;

	/**
	 * Count of notifications that are outstanding in the "wait" cycle.
	 */
	int m_notifications;

	/**
	 * Counter for generating task ids.
	 */
	protected int m_taskCounter;

	/**
	 * Connection to the task manager.
	 */
	protected IPGPushConnectorOut m_proposalInputToTaskManager;

	/**
	 * Connection to the task manager.
	 */
	protected StringPushConnectorOut m_retractionInputToTaskManager;

	/**
	 * Hashtable that stores tasks that have been proposed.
	 */
	protected Hashtable<String, InformationProcessingTask> m_proposedTasks;

	private GMResponseHandler m_responseRunnable;

	/**
	 * Constructor.
	 * 
	 * @param _id
	 *            Unique component id.
	 */
	public ManagedProcess(String _id) {
		super(_id);
		m_proposalInputToTaskManager = null;
		m_taskCounter = 0;
		m_proposedTasks = new Hashtable<String, InformationProcessingTask>();
		m_processingResultToGoalManager = null;
		m_retractionInputToTaskManager = null;
		m_taskDescriptionRegistration = null;
		m_taskNotificationLock = new Object();
		m_notifications = 0;
	}

	/**
	 * Receive a response from the task manager and call the correct
	 * result-specific method.
	 * 
	 * @param _taskID
	 *            The id of the task.
	 * @param _decision
	 *            The task manager decision.
	 */
	private void handleTaskManagerResponse(String _taskID,
			TaskManagementDecision _decision) {

		// inform sub-class of result
		if (_decision.value() == TaskManagementDecision.GOAL_ADOPTED.value()) {
			logTaskAdopted(_taskID);
			taskAdopted(_taskID);
			synchronized (m_taskNotificationLock) {
				m_notifications++;
				m_taskNotificationLock.notifyAll();
			}

		}
		else if (_decision.value() == TaskManagementDecision.GOAL_REJECTED
				.value()) {
			taskRejected(_taskID);
			synchronized (m_taskNotificationLock) {
				m_notifications++;
				m_taskNotificationLock.notifyAll();
			}

		}
		else if (_decision.value() == TaskManagementDecision.GOAL_WAITING
				.value()) {
			// just repropose for a laugh right now...

			try {
				proposeInformationProcessingTask(_taskID, m_proposedTasks
						.get(_taskID).m_taskName);
			}
			catch (SubarchitectureProcessException e) {
				e.printStackTrace();
			}
		}
		else {
			throw new RuntimeException("Unknown task manager response");
		}

	}

	/**
	 * @param _taskID
	 */
	private void logTaskAdopted(String _taskID) {
		logEvent(ComponentEventType.START, getProcessIdentifier().toString(),
				m_subarchitectureID, m_proposedTasks.get(_taskID).m_taskName,
				_taskID);
	}

	/**
	 * @param _taskID
	 */
	private void logTaskComplete(String _taskID) {

		logEvent(ComponentEventType.END, getProcessIdentifier().toString(),
				m_subarchitectureID, m_proposedTasks.get(_taskID).m_taskName,
				_taskID);
	}

	/**
	 * @param _taskID
	 * @param _name
	 */
	private void logTaskProposed(String _taskID, String _name) {
		logEvent(ComponentEventType.PROPOSED,
				getProcessIdentifier().toString(), m_subarchitectureID, _name,
				_taskID);
	}

	/**
	 * Generate a new unique task id.
	 * 
	 * @return A unique task id.
	 */
	protected String newTaskID() {
		return "IPG:" + getProcessIdentifier() + ":" + m_taskCounter++;
	}

	/**
	 * Propose the processing of a task.
	 * 
	 * @param _taskID
	 *            The id used to refer to this task.
	 * @param _taskName
	 *            The name of the task.
	 * @throws SubarchitectureProcessException
	 *             if connection to goal manager is not set or on comms error.
	 */
	protected void proposeInformationProcessingTask(String _taskID,
			String _taskName) throws SubarchitectureProcessException {

		assert (_taskID != null);
		assert (_taskName != null);

		logTaskProposed(_taskID, _taskName);

		if (m_proposalInputToTaskManager == null) {
			throw new SubarchitectureProcessException(
					"Push connection not to sa goal manager not set.");
		}

		// create the goal structure
		InformationProcessingTask ipg = new InformationProcessingTask(_taskID,
				_taskName);

		// store it as proposed
		m_proposedTasks.put(_taskID, ipg);

		// push goal to s-a goal manager
		m_proposalInputToTaskManager.push(getProcessIdentifier().toString(),
				ipg);

		debug("proposed: " + CASTUtils.toString(ipg));
	}

	/**
	 * Register a list of tasks that this component can perform.
	 * 
	 * @param _taskDesc
	 *            An array of task descriptions.
	 * @throws SubarchitectureProcessException
	 *             if connection to goal manager is not set or on comms error.
	 */
	protected void registerTaskDescriptions(TaskDescription[] _taskDesc)
			throws SubarchitectureProcessException {

		if (m_taskDescriptionRegistration == null) {
			throw new SubarchitectureProcessException(
					"Push connection not to sa goal manager not set.");
		}

		// push registration to s-a goal manager
		m_taskDescriptionRegistration.push(getProcessIdentifier().toString(),
				_taskDesc);
	}

	/**
	 * Retract the proposal of processing a task.
	 * 
	 * @param _taskID
	 *            The id of the task.
	 * @throws SubarchitectureProcessException
	 *             if connection to goal manager is not set or on comms error.
	 */
	protected void retractInformationProcessingTask(String _taskID)
			throws SubarchitectureProcessException {

		if (m_retractionInputToTaskManager == null) {
			throw new SubarchitectureProcessException(
					"Push connection not to sa goal manager not set.");
		}

		// push retraction to s-a goal manager
		m_retractionInputToTaskManager.push(getProcessIdentifier().toString(),
				_taskID);
	}

	/**
	 * Empty runComponent method. Does nothing.
	 * 
	 * @see cast.core.components.CASTComponent#runComponent()
	 */
	protected void runComponent() {
		// don't do anything if not overridden
	}

	/**
	 * Receive a signal that an information processing task has been accepted.
	 * 
	 * @param _taskID
	 */
	protected abstract void taskAdopted(String _taskID);

	/**
	 * Signal the outcome of goal processing to the sub-architecture goal
	 * manager.
	 * 
	 * @param _taskID
	 *            The task id.
	 * @param _outcome
	 *            The task outcome.
	 * @throws SubarchitectureProcessException
	 *             if connection to goal manager is not set.`
	 */
	protected void taskComplete(String _taskID, TaskOutcome _outcome)
			throws SubarchitectureProcessException {

		assert (_taskID != null);
		assert (_outcome != null);

		logTaskComplete(_taskID);

		// remove the goal from the hashtable
		m_proposedTasks.remove(_taskID);

		if (m_processingResultToGoalManager != null) {
			m_processingResultToGoalManager.push(getProcessIdentifier()
					.toString(), new TaskResult(_taskID, _outcome));
		}
		else {
			throw new SubarchitectureProcessException(
					"Goal processing result connection to goal manager has not been set up.");
		}

	}

	/**
	 * Receive a signal that an information processing goal has been rejected.
	 * 
	 * @param _taskID
	 */
	protected abstract void taskRejected(String _taskID);

	/**
	 * Put the process to sleep until a new task notification is received. This
	 * method will sleep until either the taskAdopted or taskRejected functions
	 * next <b>exit</b>, unless these methods have exited more times than this
	 * method has been called. In this case the method does not block, as the
	 * assumption is that the process has to process these outstanding
	 * notifications first. This latter behaviour was added to fixed bug #31.
	 * 
	 * @remark Interface change: Used to take a single argument. This was
	 *         removed to simplify behaviour.
	 */
	protected void waitForNotifications() {

		synchronized (m_taskNotificationLock) {
			if (m_notifications == 0) {
				// println("blocking for new");
				try {
					m_taskNotificationLock.wait();
				}
				catch (InterruptedException e) {
					e.printStackTrace();
				}
				// the count will have been incremented in the same action as
				// the broadcast
				m_notifications--;
				assert (m_notifications >= 0);

			}
			// else decrement the coutn
			else {
				m_notifications--;
				assert (m_notifications >= 0);
			}
		}

	}

	/**
	 * Receives the GoalManagementResult from the sub-architecture goal manager
	 * and forwards this change to the underlying class in a new thread. Also
	 * locks access to the receiving thread whilst doing this.
	 * 
	 * @param _src
	 *            The component that sent the data.
	 * @param _data
	 *            Task management result.
	 * @see cast.core.interfaces.GMRPushInterface.GMRPushReceiver#receivePushData(java.lang.String,
	 *      cast.corba.autogen.CAAT.TaskManagementResult)
	 */
	public void receivePushData(String _src, TaskManagementResult _data) {

		debug("received goal management decision: " + _data.m_id);

		// if this goal was proposed by this component
		// hashtable is synchronised, so no need to protect it
		if (m_proposedTasks.containsKey(_data.m_id)) {

			if (m_responseRunnable == null) {
				m_responseRunnable = new GMResponseHandler(this);
				m_responseRunnable.start();
				new Thread(m_responseRunnable).start();
			}

			m_responseRunnable.queue(_data);

		}
	}

	/**
	 * Set push connector to task manager for task results.
	 * 
	 * @param _connectionID
	 *            The connector id.
	 * @param _out
	 *            The connector.
	 * @see cast.core.interfaces.GPRPushInterface.GPRPushSender#setPushConnector(java.lang.String,
	 *      cast.core.interfaces.GPRPushInterface.GPRPushConnectorOut)
	 */
	public void setPushConnector(String _connectionID, GPRPushConnectorOut _out) {
		m_processingResultToGoalManager = _out;
	}

	/**
	 * Set push connector to task manager for task proposition.
	 * 
	 * @param _connectionID
	 *            The connector id.
	 * @param _out
	 *            The connector.
	 * @see cast.core.interfaces.IPGPushInterface.IPGPushSender#setPushConnector(java.lang.String,
	 *      cast.core.interfaces.IPGPushInterface.IPGPushConnectorOut)
	 */
	public void setPushConnector(String _connectionID, IPGPushConnectorOut _out) {
		m_proposalInputToTaskManager = _out;
	}

	/**
	 * Set push connector to task manager for task retraction.
	 * 
	 * @param _connectionID
	 *            The connector id.
	 * @param _out
	 *            The connector.
	 * @see balt.core.connectors.push.nonprimitive.interfaces.StringPushInterface.StringPushSender#setPushConnector(java.lang.String,
	 *      balt.core.connectors.push.nonprimitive.interfaces.StringPushInterface.StringPushConnectorOut)
	 */
	public void setPushConnector(String _connectionID,
			StringPushConnectorOut _out) {
		m_retractionInputToTaskManager = _out;
	}

	/**
	 * Set push connector to task manager for task registration.
	 * 
	 * @param _connectionID
	 *            The connector id.
	 * @param _out
	 *            The connector.
	 * @see cast.core.interfaces.TDLPushInterface.TDLPushSender#setPushConnector(java.lang.String,
	 *      cast.core.interfaces.TDLPushInterface.TDLPushConnectorOut)
	 */
	public void setPushConnector(String _connectionID, TDLPushConnectorOut _out) {
		m_taskDescriptionRegistration = _out;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see caat.architecture.abstr.WorkingMemoryReaderProcess#stop()
	 */
	@Override
	public void stop() {
		super.stop();

		if (m_responseRunnable != null) {
			m_responseRunnable.stop();
		}

		synchronized (m_taskNotificationLock) {
			m_taskNotificationLock.notifyAll();
		}
	}

}
